### PR TITLE
Workflow-service: Parameters as jsonschema

### PR DESCRIPTION
**Pull request by eloycoto**
This commit introduces a new way to consume WorkFlow Parameters and
TaskParameters, so user can define custom validations and extend the Parodos ui
forms with extra information. Workflow definitions will be the same, but
builders has now a jsonschema information, like this:

```
WorkFlowTaskParameter
    .builder()
    .key("user-id")
    .description("The user id")
    .type(WorkFlowTaskParameterType.TEXT)
    .optional(false)
    .JsonSchemaOptions(Map.of(
        "minLength", "1",
        "maxLength", "64"))
    .build());
```

The controller will export the data as a Map, and with jsonschema default
information:

```

{
    "id": "778267a0-6729-4a3c-a97e-6785b377f36e",
    "name": "simpleSequentialWorkFlow_INFRASTRUCTURE_WORKFLOW",
    "type": "INFRASTRUCTURE",
    "processingType": "SEQUENTIAL",
    "author": null,
    "createDate": "2023-03-21T11:52:11.588+00:00",
    "modifyDate": "2023-03-21T11:52:11.588+00:00",
    "works": [
      {
        "id": "61b443e9-b746-4ac9-872f-ab5063c1db85",
        "name": "restCallTask",
        "workType": "TASK",
        "parameters": {
          "payload": {
            "format": "text",
            "description": "Json of what to provide for data. (ie: 'Hello!')",
            "type": "string",
            "required": "true"
          },
          "user-id": {
            "format": "text",
            "description": "The user id",
            "type": "string",
            "required": "true"
          },
          "url": {
            "format": "url",
            "description": "The Url of the service (ie: https://httpbin.org/post",
            "type": "string",
            "required": "true"
          }
        },
        "outputs": [
          "HTTP2XX",
          "OTHER"
        ]
      },
      {
        "id": "71b3f1de-e801-4aa2-936e-419ce93704bc",
        "name": "loggingTask",
        "workType": "TASK",
        "parameters": {
          "user-id": {
            "minLength": "1",
            "format": "text",
            "description": "The user id",
            "type": "string",
            "required": "true",
            "maxLength": "64"
          },
          "api-server": {
            "format": "url",
            "description": "The api server",
            "type": "string",
            "required": "true"
          }
        },
        "outputs": [
          "OTHER"
        ]
      }
    ]
}
```

This will allow users to have better integration with end users.

Fix [FLPATH-193](https://issues.redhat.com//browse/FLPATH-193)

### CODE REVIEW

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/WorkFlowTask.java [hunk 1]
```diff
@@ -52,4 +58,26 @@ default List<WorkFlowTaskOutput> getWorkFlowTaskOutputs() {
...
 	}
 
+	default HashMap<String, Map<String, String>> getAsJsonSchema() {
+		HashMap<String, Map<String, String>> result = new HashMap<>();
+		taskParamsLoop: for (WorkFlowTaskParameter workFlowTaskParameter : this.getWorkFlowTaskParameters()) {
+			if (workFlowTaskParameter == null) {
+				continue;
+			}
+
+			if (workFlowTaskParameter.getType() == null) {
```
Comment (pkliczewski) [inline parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/WorkFlowTask.java:68]: is this possible not to have a type? Sounds like a bug to me.

File: workflow-examples/src/main/java/com/redhat/parodos/examples/simple/task/RestAPIWorkFlowTask.java [hunk 2]
```diff
@@ -73,7 +73,7 @@ public List<WorkFlowTaskParameter> getWorkFlowTaskParameters() {
 						.type(WorkFlowTaskParameterType.URL).build(),
 				WorkFlowTaskParameter.builder().key(PAYLOAD_KEY)
 						.description("Json of what to provide for data. (ie: 'Hello!')").optional(false)
-						.type(WorkFlowTaskParameterType.PASSWORD).build(),
+						.type(WorkFlowTaskParameterType.TEXT).build(),
```
Comment (pkliczewski) [inline workflow-examples/src/main/java/com/redhat/parodos/examples/simple/task/RestAPIWorkFlowTask.java:76]: What is the reason to change type here?

File: workflow-service/src/main/java/com/redhat/parodos/workflow/definition/service/WorkFlowDefinitionServiceImpl.java [hunk 3]
```diff
@@ -153,17 +159,15 @@ public WorkFlowDefinitionResponseDTO getWorkFlowDefinitionById(UUID id) {
...
-				.processingType(workFlowDefinition.getProcessingType())
+				.parameterFromString(workFlowDefinition.getParameters()).author(workFlowDefinition.getAuthor())
+				.createDate(workFlowDefinition.getCreateDate()).modifyDate(workFlowDefinition.getModifyDate())
+				.type(workFlowDefinition.getType().toString()).processingType(workFlowDefinition.getProcessingType())
 				.works(buildWorkFlowWorksDTOs(workFlowDefinition, workFlowWorkDependencies)).build();
 	}
 
 	@Override
 	public WorkFlowDefinitionResponseDTO getWorkFlowDefinitionByName(String name) {
+		// ELOY
```
Comment (pkliczewski) [inline workflow-service/src/main/java/com/redhat/parodos/workflow/definition/service/WorkFlowDefinitionServiceImpl.java:170]: Do we need this comment?

File: workflow-examples/src/main/java/com/redhat/parodos/examples/simple/task/RestAPIWorkFlowTask.java (hunk 2 above)
Comment (eloycoto) [reply workflow-examples/src/main/java/com/redhat/parodos/examples/simple/task/RestAPIWorkFlowTask.java:76]: json payload is not a password type at all.

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/WorkFlowTask.java (hunk 1 above)
Comment (eloycoto) [reply parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/WorkFlowTask.java:68]: likely not, but the check does not harm, at the end is something submitted by a user

File: parodos-model-api/src/test/java/com/redhat/parodos/workflow/parameter/WorkFlowParameterTypeTest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.List;
+import java.util.Map;
+
+public class WorkFlowParameterTypeTest extends TestCase {
+
+	@Test
+	public void testJsonSchema() {
+		WorkFlowParameterType[] inputValues = { WorkFlowParameterType.PASSWORD, WorkFlowParameterType.TEXT,
+				WorkFlowParameterType.EMAIL, WorkFlowParameterType.NUMBER, WorkFlowParameterType.URL,
+				WorkFlowParameterType.DATE };
```
Comment (gciavarrini) [inline parodos-model-api/src/test/java/com/redhat/parodos/workflow/parameter/WorkFlowParameterTypeTest.java:18]: wdyt to improve the formatting?

File: workflow-service/src/main/java/com/redhat/parodos/workflow/definition/service/WorkFlowDefinitionServiceImpl.java [hunk 5]
```diff
@@ -82,15 +82,28 @@ public WorkFlowDefinitionServiceImpl(WorkFlowDefinitionRepository workFlowDefini
...
 		// set and save workflow definition
-		WorkFlowDefinition workFlowDefinition = workFlowDefinitionRepository
-				.save(WorkFlowDefinition.builder().name(workFlowName).type(workFlowType).createDate(new Date())
-						.parameters(WorkFlowDTOUtil.writeObjectValueAsString(workFlowParameters)).modifyDate(new Date())
-						.numberOfWorks(works.size()).processingType(workFlowProcessingType.name()).build());
+		WorkFlowDefinition workFlowDefinition = workFlowDefinitionRepository.save(WorkFlowDefinition.builder()
+				.name(workFlowName).type(workFlowType).createDate(new Date())
+				.parameters(WorkFlowDTOUtil.writeObjectValueAsString(convertWorkFlowParameters(workFlowParameters)))
+				.modifyDate(new Date()).numberOfWorks(works.size()).processingType(workFlowProcessingType.name())
+				.build());
```
Comment (gciavarrini) [inline workflow-service/src/main/java/com/redhat/parodos/workflow/definition/service/WorkFlowDefinitionServiceImpl.java:106]: Difficult to read: can you fix it?

This also apply to other similar code blocks.

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/WorkFlowTask.java [hunk 6]
```diff
@@ -15,12 +15,18 @@
...
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+
 import com.redhat.parodos.workflow.task.parameter.WorkFlowTaskParameter;
+import com.redhat.parodos.workflow.task.parameter.WorkFlowTaskParameterType;
 import com.redhat.parodos.workflows.work.Work;
 import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
```
Comment (gciavarrini) [inline parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/WorkFlowTask.java:30]: Unsed import

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/WorkFlowTask.java [hunk 7]
```diff
@@ -52,4 +58,26 @@ default List<WorkFlowTaskOutput> getWorkFlowTaskOutputs() {
 
 	}
 
+	default HashMap<String, Map<String, String>> getAsJsonSchema() {
+		HashMap<String, Map<String, String>> result = new HashMap<>();
+		taskParamsLoop: for (WorkFlowTaskParameter workFlowTaskParameter : this.getWorkFlowTaskParameters()) {
```
Comment (gciavarrini) [inline parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/WorkFlowTask.java:63]: Is the label `taskParamsLoop` necessary?

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/WorkFlowTask.java [hunk 8]
```diff
@@ -52,4 +58,26 @@ default List<WorkFlowTaskOutput> getWorkFlowTaskOutputs() {
...
+
+			if (workFlowTaskParameter.getType() == null) {
+				continue;
+			}
+
+			Map<String, String> properties = workFlowTaskParameter.getType().getAsJsonSchema();
+			properties.put("required", String.format("%s", !workFlowTaskParameter.isOptional()));
+			properties.put("description", workFlowTaskParameter.getDescription());
+			if (workFlowTaskParameter.getJsonSchemaOptions() != null) {
+				workFlowTaskParameter.getJsonSchemaOptions().forEach(properties::put);
```
Comment (gciavarrini) [inline parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/WorkFlowTask.java:76]: can you check if it's possible to simplify this in `getJsonSchemaOptions().putAll(properties);`?

**Review by gciavarrini — CHANGES_REQUESTED** [review]
Overall lgtm, just some suggestion here and there :)

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/parameter/WorkFlowTaskParameterType.java [hunk 9]
```diff
@@ -24,6 +27,36 @@
...
+				break;
+			case URL:
+				properties.put("type", "string");
+				properties.put("format", "url");
+				break;
+			case DATE:
+				properties.put("type", "string");
+				properties.put("format", "date");
+				break;
+		}
```
[the remark this answers is not in the document]
Comment (eloycoto) [reply parodos-model-api/src/main/java/com/redhat/parodos/workflow/task/parameter/WorkFlowTaskParameterType.java:60]: addressed

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/parameter/WorkFlowParameterType.java [hunk 10]
```diff
@@ -23,6 +26,36 @@
...
+				break;
+			case URL:
+				properties.put("type", "string");
+				properties.put("format", "url");
+				break;
+			case DATE:
+				properties.put("type", "string");
+				properties.put("format", "date");
+				break;
+		}
```
[the remark this answers is not in the document]
Comment (eloycoto) [reply parodos-model-api/src/main/java/com/redhat/parodos/workflow/parameter/WorkFlowParameterType.java:59]: addressed

**Pull request merged**